crates/sui-storage/src/lock_service.rs

```diff
@@ -181,17 +183,31 @@ impl LockServiceImpl {
 
     /// Gets ObjectLockInfo that represents state of lock on an object.
     /// Returns SuiError::ObjectNotFound if cannot find lock record for this object
-    fn get_lock(&self, obj_ref: ObjectRef) -> SuiLockResult {
+    fn get_lock(&self, obj_ref: ObjectRef, epoch_id: EpochId) -> SuiLockResult {
         Ok(
             if let Some(lock_info) = self
                 .transaction_lock
                 .get(&obj_ref)
                 .map_err(SuiError::StorageError)?
             {
                 match lock_info {
-                    Some(lock_info) => ObjectLockStatus::LockedToTx {
-                        locked_by_tx: lock_info,
-                    },
+                    Some(lock_info) => {
+                        match Ord::cmp(&lock_info.epoch, &epoch_id) {
+                            // If the object was locked in a previous epoch, we can say that it's
+                            // no longer locked and is considered as just Initialized.
+                            Ordering::Less => ObjectLockStatus::Initialized,
+                            Ordering::Equal => ObjectLockStatus::LockedToTx {
+                                locked_by_tx: lock_info,
+                            },
+                            Ordering::Greater => {
+                                return Err(SuiError::ObjectLockedAtFutureEpoch {
+                                    obj_refs: vec![obj_ref],
```

Comment:
nit: is it possible to add the tx which is holding the lock in future epoch as well here (might be useful for debugging)